File: python/x402/mechanisms/evm/batch_settlement/client/recovery.py
"""Corrective-402 recovery flows."""

from __future__ import annotations

try:
    from eth_account import Account
    from eth_account.messages import encode_typed_data
    from eth_utils import to_checksum_address
except ImportError as e:
    raise ImportError(
        "EVM mechanism requires ethereum packages. Install with: pip install x402[evm]"
    ) from e

from .....schemas import PaymentRequired, PaymentRequirements
from ...utils import get_evm_chain_id
from ..constants import SCHEME_BATCH_SETTLEMENT, VOUCHER_TYPES
from ..errors import ERR_CUMULATIVE_AMOUNT_BELOW_CLAIMED, ERR_CUMULATIVE_AMOUNT_MISMATCH
from ..types import ChannelStateExtra, VoucherStateExtra
from ..utils import (
    coerce_bytes32,
    compute_channel_id,
    get_batch_settlement_eip712_domain,
)
from .channel import (
    BatchSettlementClientDeps,
    build_channel_config,
    read_channel_balance_and_total_claimed,
)
from .storage import BatchSettlementClientContext


def process_corrective_payment_required(
    deps: BatchSettlementClientDeps,
    payment_required: PaymentRequired,
) -> bool:
    """Handle a corrective 402 by resyncing local channel state."""
    err = payment_required.error
    if err not in (ERR_CUMULATIVE_AMOUNT_MISMATCH, ERR_CUMULATIVE_AMOUNT_BELOW_CLAIMED):
        return False

    accept: PaymentRequirements | None = None
    for a in payment_required.accepts:
        if a.scheme == SCHEME_BATCH_SETTLEMENT:
            accept = a
            break
    if accept is None:
        return False

    extra = accept.extra or {}
    cs_raw = extra.get("channelState")
    vs_raw = extra.get("voucherState")
    channel_state = ChannelStateExtra.from_dict(cs_raw) if isinstance(cs_raw, dict) else None
    voucher_state = VoucherStateExtra.from_dict(vs_raw) if isinstance(vs_raw, dict) else None

    has_sig = (
        channel_state is not None
        and channel_state.charged_cumulative_amount is not None
        and voucher_state is not None
        and voucher_state.signed_max_claimable is not None
        and voucher_state.signature is not None
    )

    if not has_sig:
        return recover_from_onchain_state(deps, accept)

    assert channel_state is not None and voucher_state is not None
    return recover_from_signature(deps, accept, channel_state, voucher_state)


def recover_from_signature(
    deps: BatchSettlementClientDeps,
    accept: PaymentRequirements,
    channel_state: ChannelStateExtra,
    voucher_state: VoucherStateExtra,
) -> bool:
    """Recover when the corrective 402 includes a server-stored voucher signature."""
    if (
        channel_state.charged_cumulative_amount is None
        or voucher_state.signed_max_claimable is None
        or voucher_state.signature is None
    ):
        return False

    charged = int(channel_state.charged_cumulative_amount)
    signed = int(voucher_state.signed_max_claimable)
    sig = voucher_state.signature

    if charged > signed:
        return False

    config = build_channel_config(deps, accept)
    channel_id = compute_channel_id(config, str(accept.network))

    if not hasattr(deps.signer, "read_contract"):
        return False
    ch_balance, ch_total_claimed = read_channel_balance_and_total_claimed(deps.signer, channel_id)

    if charged < ch_total_claimed:
        return False

    chain_id = get_evm_chain_id(str(accept.network))
    domain = get_batch_settlement_eip712_domain(chain_id)
    message = {
        "channelId": coerce_bytes32(channel_id),
        "maxClaimableAmount": signed,
    }

    signable = encode_typed_data(
        domain_data=domain,
        message_types=VOUCHER_TYPES,
        message_data=message,
    )
    sig_bytes = bytes.fromhex(sig.removeprefix("0x"))
    recovered = Account.recover_message(signable, signature=sig_bytes)

    if deps.payer_authorizer is not None:
        expected = to_checksum_address(deps.payer_authorizer)
    elif deps.voucher_signer is not None:
        expected = to_checksum_address(deps.voucher_signer.address)
    else:
        expected = to_checksum_address(deps.signer.address)

    if to_checksum_address(recovered) != expected:
        return False

    ctx = BatchSettlementClientContext(
        charged_cumulative_amount=str(charged),
        signed_max_claimable=str(signed),
        signature=sig,
        balance=str(ch_balance),
        total_claimed=str(ch_total_claimed),
    )
    deps.storage.set(channel_id.lower(), ctx)
    return True


def recover_from_onchain_state(
    deps: BatchSettlementClientDeps,
    accept: PaymentRequirements,
) -> bool:
    """Recover channel state purely from on-chain `channels(channelId)`."""
    if not hasattr(deps.signer, "read_contract"):
        return False
    config = build_channel_config(deps, accept)
    channel_id = compute_channel_id(config, str(accept.network))
    ch_balance, ch_total_claimed = read_channel_balance_and_total_claimed(deps.signer, channel_id)
    ctx = BatchSettlementClientContext(
        charged_cumulative_amount=str(ch_total_claimed),
        balance=str(ch_balance),
        total_claimed=str(ch_total_claimed),
    )
    deps.storage.set(channel_id.lower(), ctx)
    return True


__all__ = [
    "process_corrective_payment_required",
    "recover_from_signature",
    "recover_from_onchain_state",
]

File: typescript/packages/mechanisms/evm/src/batch-settlement/client/recovery.ts
import type { PaymentRequired, PaymentRequirements } from "@x402/core/types";
import { getAddress, recoverTypedDataAddress } from "viem";
import { BATCH_SETTLEMENT_SCHEME, voucherTypes } from "../constants";
import type { BatchSettlementClientContext } from "./storage";
import { computeChannelId, getBatchSettlementEip712Domain } from "../utils";
import { getEvmChainId } from "../../utils";
import {
  type BatchSettlementClientDeps,
  buildChannelConfig,
  readChannelBalanceAndTotalClaimed,
} from "./channel";
import * as Errors from "../errors";
import type { BatchSettlementChannelStateExtra, BatchSettlementVoucherStateExtra } from "../types";

/**
 * Handles a corrective 402 response from the server when the client's
 * cumulative base is out of sync.
 *
 * Validates the server-provided state (chargedCumulativeAmount,
 * signedMaxClaimable, signature) against onchain data and the client's own
 * signing key, then updates the local channel state if everything checks out.
 *
 * @param deps - Signer + storage + identity inputs.
 * @param paymentRequired - The decoded 402 response body.
 * @returns `true` if the channel state was successfully resynced and the request can be retried.
 */
export async function processCorrectivePaymentRequired(
  deps: BatchSettlementClientDeps,
  paymentRequired: PaymentRequired,
): Promise<boolean> {
  if (
    paymentRequired.error !== Errors.ErrCumulativeAmountMismatch &&
    paymentRequired.error !== Errors.ErrCumulativeAmountBelowClaimed
  ) {
    return false;
  }

  const accept = paymentRequired.accepts.find(a => a.scheme === BATCH_SETTLEMENT_SCHEME);
  if (!accept) {
    return false;
  }

  const channelState = accept.extra.channelState as BatchSettlementChannelStateExtra | undefined;
  const voucherState = accept.extra.voucherState as BatchSettlementVoucherStateExtra | undefined;
  const hasSig =
    channelState?.chargedCumulativeAmount !== undefined &&
    voucherState?.signedMaxClaimable !== undefined &&
    voucherState.signature !== undefined;

  if (!hasSig) {
    return recoverFromOnChainState(deps, accept);
  }

  return recoverFromSignature(deps, accept, channelState, voucherState);
}

/**
 * Recovers channel state from a corrective 402 that includes a server-provided
 * voucher signature. Verifies the signature matches the client's own signing
 * key before accepting.
 *
 * @param deps - Signer + storage + identity inputs.
 * @param accept - Batch settlement payment requirements from the corrective 402.
 * @param channelState - Server channel snapshot from `accept.extra.channelState`.
 * @param voucherState - Latest signed voucher proof from `accept.extra.voucherState`.
 * @returns `true` when local channel state was updated successfully.
 */
export async function recoverFromSignature(
  deps: BatchSettlementClientDeps,
  accept: PaymentRequirements,
  channelState: BatchSettlementChannelStateExtra,
  voucherState: BatchSettlementVoucherStateExtra,
): Promise<boolean> {
  const chargedRaw = channelState.chargedCumulativeAmount;
  const signedRaw = voucherState.signedMaxClaimable;
  const sig = voucherState.signature as `0x${string}`;

  const charged = BigInt(String(chargedRaw));
  const signed = BigInt(String(signedRaw));

  if (charged > signed) {
    return false;
  }

  const config = buildChannelConfig(deps, accept);
  const channelId = computeChannelId(config, accept.network);

  if (!deps.signer.readContract) {
    return false;
  }

  const [chBalance, chTotalClaimed] = await readChannelBalanceAndTotalClaimed(
    deps.signer,
    channelId,
  );

  if (charged < chTotalClaimed) {
    return false;
  }

  const chainId = getEvmChainId(accept.network);
  const recovered = await recoverTypedDataAddress({
    domain: getBatchSettlementEip712Domain(chainId),
    types: voucherTypes,
    primaryType: "Voucher",
    message: {
      channelId,
      maxClaimableAmount: signed,
    },
    signature: sig,
  });

  const expectedSigner = getAddress(
    deps.payerAuthorizer ?? deps.voucherSigner?.address ?? deps.signer.address,
  );
  if (recovered.toLowerCase() !== expectedSigner.toLowerCase()) {
    return false;
  }

  const ctx: BatchSettlementClientContext = {
    chargedCumulativeAmount: charged.toString(),
    signedMaxClaimable: signed.toString(),
    signature: sig,
    balance: chBalance.toString(),
    totalClaimed: chTotalClaimed.toString(),
  };

  await deps.storage.set(channelId.toLowerCase(), ctx);
  return true;
}

/**
 * Recovers channel state purely from onchain state when the server has no stored
 * voucher (e.g. after a cooperative refund deleted the channel record). The onchain
 * `totalClaimed` becomes the new baseline — no signature verification is
 * needed because the contract is the source of truth when no outstanding
 * voucher exists.
 *
 * @param deps - Signer + storage + identity inputs.
 * @param accept - Batch settlement payment requirements from the corrective 402.
 * @returns `true` when local channel state was updated from onchain data.
 */
export async function recoverFromOnChainState(
  deps: BatchSettlementClientDeps,
  accept: PaymentRequirements,
): Promise<boolean> {
  if (!deps.signer.readContract) {
    return false;
  }

  const config = buildChannelConfig(deps, accept);
  const channelId = computeChannelId(config, accept.network);

  const [chBalance, chTotalClaimed] = await readChannelBalanceAndTotalClaimed(
    deps.signer,
    channelId,
  );

  const ctx: BatchSettlementClientContext = {
    chargedCumulativeAmount: chTotalClaimed.toString(),
    balance: chBalance.toString(),
    totalClaimed: chTotalClaimed.toString(),
  };

  await deps.storage.set(channelId.toLowerCase(), ctx);
  return true;
}
